src/test/rbd_mirror/test_mock_MirrorStatusUpdater.cc

```diff
@@ -385,6 +417,69 @@ TEST_F(TestMockMirrorStatusUpdater, OverwriteStatus) {
                                   *mock_mirror_status_watcher);
 }
 
+TEST_F(TestMockMirrorStatusUpdater, RemoveStatus) {
+  MockMirrorStatusUpdater mock_mirror_status_updater(m_local_io_ctx,
+                                                     m_mock_threads, "");
+  MockMirrorStatusWatcher* mock_mirror_status_watcher =
+    new MockMirrorStatusWatcher();
+
+  InSequence seq;
+
+  Context* timer_event = nullptr;
+  init_mirror_status_updater(mock_mirror_status_updater,
+                             *mock_mirror_status_watcher, &timer_event);
+
+  C_SaferCond ctx;
+  mock_mirror_status_updater.set_mirror_image_status("1", {}, false);
+  expect_work_queue(false);
+  mock_mirror_status_updater.remove_mirror_image_status("1", false, &ctx);
+  ASSERT_EQ(0, ctx.wait());
+
+  Context* update_task = nullptr;
+  fire_timer_event(&timer_event, &update_task);
+
+  expect_mirror_status_remove({"1"}, 0);
+  update_task->complete(0);
```

Comment:
For some reasons the mocked methods in `expect_mirror_status_remove` are called after this line and it break this test sometimes. I don't know if it's a mock problem or something missing in the rbd_cls code... Both looks good to me but there should be a catch...

Reply:
It is because `update_task->complete(0)` calls `MirrorStatusUpdater::update_task` but it does not wait for `handle_update_task` to complete, so the test may call `shut_down_mirror_status_updater` before `handle_update_task` is called.

I suppose you should wait for remove to complete before calling shutdown,  similarly to how it waits in `TestMockMirrorStatusUpdater.RemoveRefreshInFlightStatus`.

Reply:
Thanks should be fixed now!